src/statmech.jl

```diff
@@ -0,0 +1,193 @@
+    for (i, a) in enumerate(atoms)
+        var_indices[a] = i
+    end
+
+    #the residuals of the molecular equilibrium equations parametrized by T, electron number density
+    #and number densities of each element [cm^-3]
+    function system(T, nₜ, nₑ)
+        #`residuals!` puts the residuals the system of molecular equilibrium equations in `F`
+        #`x` is a vector containing the number density of the neutral species of each element
+        function residuals!(F, x)
+            for (i, a) in enumerate(atoms)
+                #LHS: total number of atoms
+                F[i] = nₜ * absolute_abundances[a]
+                wII, wIII = saha_ion_weights(T, nₑ, a, ionization_energies, partition_fns)
+                #RHS: first through third ionization states
+                F[i] -= (1 + wII + wIII) * x[i]
+            end
+            for m in molecules
+                el1, el2 = get_atoms(m)
+                i1 = var_indices[el1]
+                i2 = var_indices[el2]
+                nₘ = x[i1] * x[i2] * kboltz_cgs * T / 10^equilibrium_constants[m](T)
+                #RHS: subtract atoms which are part of mollecules
+                F[i1] -= nₘ
+                F[i2] -= nₘ
+            end
+        end
+    end
+
+    #passing atoms and molecules might seem a little weird architecturally, but it's partly in 
+    #anticipation if automatically culling the list of species considered by what's in the line list
+    #in the future
+    (atoms=atoms, molecules=molecules, equations=system, absolute_abundances,
+     ionization_energies=ionization_energies, partition_fns=partition_fns, 
+     equilibrium_constants=equilibrium_constants)
+end
+
+"""
+Iteratively solve for the number density of each species. Returns a Dict mapping species to number 
+densities.
+
+arguments:
+- the system of molecular equilibrium equations `MEQs` (the thing returned by 
+`molecular_equilibrium_equations`)
+- the temperature `T`, 
+- the number density of non-electron particles `nₜ`
+- the electron number density `nₑ`
+- a Dict of N_X/N_total, `absolute_abundances`
+- a Dict of ionization energies, `ionization_energies`
+- a Dict of partition functions, `partition_fns`
+- a Dict of log molecular equilibrium constants, `equilibrium_constants`, in partial pressure form.
+"""
+function molecular_equilibrium(MEQs, T, nₜ, nₑ;
+                               x0=[nₜ*MEQs.absolute_abundances[a]* 0.8 for a in MEQs.atoms]) :: Dict
+    #numerically solve for equlibrium.  This uses finite difference rather that autodiff bacause 
+    #it's fast enough this way and requires fewer deps, but enabling autodiff is trivial
+    sol = nlsolve(MEQs.equations(T, nₜ, nₑ), x0; iterations=20, store_trace=true, ftol=nₜ * 1e-12, 
+                  autodiff=:forward)
+    if !sol.f_converged
+        error("Mollecular equlibrium unconverged", sol, "\n", sol.trace)
+    end
+
+    #start with the neutral atomic species
+    number_densities = Dict((MEQs.atoms .* "_I") .=> sol.zero)
```

Comment:
So the the `*` operator is used for string concatenation in Julia instead of `+`?

Reply:
Yup. I think people prefer it because string concatenation non-commutative, and some “types of multiplication” (e.g. matrix multiplication) aren’t commutative, but addition pretty much always is. Also multiplication is notated by concatenating symbols in math, so it’s a nice parallel.